### ISSUE TITLE
Fail to install on EC2 G5XL Deep Learning AMI, PyTc 1.12.1 Cuda 11.6

### ISSUE THREAD

**Issue by eran-sefirot** [facebookresearch/xformers#542]
# 🐛 Bug

Fail to install on EC2 G5XL Deep Learning AMI, PyTc 1.12.1 Cuda 11.6

Fail to install on EC2 G5XL Deep Learning AMI, PyTc 1.12.1 Cuda 11.6

I am trying to install SD 2.0 which is not loading on this instance size, probably since it's short on RAM. (16GB) hoping xformers would enable us to use this cost effective instance type.

However installation fails with the stack below.

## Command
run the following:

source activate pytorch

pip install ninja

pip install -v -U git+https://github.com/facebookresearch/xformers.git@main#egg=xformers

error meeage:
× Encountered error while trying to install package.
╰─> xformers

see stack below:

  ninja: build stopped: subcommand failed.
  Traceback (most recent call last):
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 1808, in _run_ninja_build
      subprocess.run(
    File "/opt/conda/envs/pytorch/lib/python3.9/subprocess.py", line 528, in run
      raise CalledProcessError(retcode, process.args,
  subprocess.CalledProcessError: Command '['ninja', '-v']' returned non-zero exit status 1.

  The above exception was the direct cause of the following exception:

  Traceback (most recent call last):
    File "<string>", line 2, in <module>
    File "<pip-setuptools-caller>", line 34, in <module>
    File "/tmp/pip-install-p9onxlpn/xformers_fd982e89e4a14655a3d3d12b9c744cbd/setup.py", line 264, in <module>
      setuptools.setup(
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/__init__.py", line 87, in setup
      return distutils.core.setup(**attrs)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 185, in setup
      return run_commands(dist)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 201, in run_commands
      dist.run_commands()
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 968, in run_commands
      self.run_command(cmd)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/dist.py", line 1217, in run_command
      super().run_command(command)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 987, in run_command
      cmd_obj.run()
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/command/install.py", line 68, in run
      return orig.install.run(self)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/command/install.py", line 698, in run
      self.run_command('build')
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/cmd.py", line 319, in run_command
      self.distribution.run_command(command)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/dist.py", line 1217, in run_command
      super().run_command(command)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 987, in run_command
      cmd_obj.run()
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/command/build.py", line 132, in run
      self.run_command(cmd_name)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/cmd.py", line 319, in run_command
      self.distribution.run_command(command)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/dist.py", line 1217, in run_command
      super().run_command(command)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 987, in run_command
      cmd_obj.run()
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/command/build_ext.py", line 84, in run
      _build_ext.run(self)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/command/build_ext.py", line 346, in run
      self.build_extensions()
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 765, in build_extensions
      build_ext.build_extensions(self)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/command/build_ext.py", line 466, in build_extensions
      self._build_extensions_serial()
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/command/build_ext.py", line 492, in _build_extensions_serial
      self.build_extension(ext)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/command/build_ext.py", line 246, in build_extension
      _build_ext.build_extension(self, ext)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/command/build_ext.py", line 547, in build_extension
      objects = self.compiler.compile(
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 586, in unix_wrap_ninja_compile
      _write_ninja_file_and_compile_objects(
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 1487, in _write_ninja_file_and_compile_objects
      _run_ninja_build(
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 1824, in _run_ninja_build
      raise RuntimeError(message) from e
  RuntimeError: Error compiling objects for extension
  error: subprocess-exited-with-error
  
  × Running setup.py install for xformers did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /opt/conda/envs/pytorch/bin/python3.9 -u -c '
  exec(compile('"'"''"'"''"'"'
  # This is <pip-setuptools-caller> -- a caller that pip uses to run setup.py
  #
  # - It imports setuptools before invoking setup.py, to enable projects that directly
  #   import from `distutils.core` to work with newer packaging standards.
  # - It provides a clear error message when setuptools is not installed.
  # - It sets `sys.argv[0]` to the underlying `setup.py`, when invoking `setup.py` so
  #   setuptools doesn'"'"'t think the script is `-c`. This avoids the following warning:
  #     manifest_maker: standard file '"'"'-c'"'"' not found".
  # - It generates a shim setup.py, for handling setup.cfg-only projects.
  import os, sys, tokenize
  
  try:
      import setuptools
  except ImportError as error:
      print(
          "ERROR: Can not execute `setup.py` since setuptools is not available in "
          "the build environment.",
          file=sys.stderr,
      )
      sys.exit(1)
  
  __file__ = %r
  sys.argv[0] = __file__
  
  if os.path.exists(__file__):
      filename = __file__
      with tokenize.open(__file__) as f:
          setup_py_code = f.read()
  else:
      filename = "<auto-generated setuptools caller>"
      setup_py_code = "from setuptools import setup; setup()"
  
  exec(compile(setup_py_code, filename, "exec"))
  '"'"''"'"''"'"' % ('"'"'/tmp/pip-install-p9onxlpn/xformers_fd982e89e4a14655a3d3d12b9c744cbd/setup.py'"'"',), "<pip-setuptools-caller>", "exec"))' install --record /tmp/pip-record-bxhl5nt3/install-record.txt --single-version-externally-managed --compile --install-headers /opt/conda/envs/pytorch/include/python3.9/xformers
  cwd: /tmp/pip-install-p9onxlpn/xformers_fd982e89e4a14655a3d3d12b9c744cbd/
  Running setup.py install for xformers ... error
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> xformers
## Environment

Please copy and paste the output from the
environment collection script from PyTorch
(or fill out the checklist below manually).

You can run the script with:

```bash
# For security purposes, please check the contents of collect_env.py before running it.
python -m torch.utils.collect_env
```

- PyTorch Version (e.g., 1.0):
- OS (e.g., Linux):
- How you installed PyTorch (`conda`, `pip`, source):
- Build command you used (if compiling from source):
- Python version:
- CUDA/cuDNN version:
- GPU models and configuration:
- Any other relevant information:

## Additional context

We're trying to build a VR engine on top of SD:
studio.sefirot.io

**Comment by danthe3rd** [comment]
Hi,

It's possible that the compilation might be OOMing. Can you try to reduce "MAX_JOBS", as written in the [troubleshooting section](https://github.com/facebookresearch/xformers#install-troubleshooting) ?

**Comment by eran-sefirot** [comment]
OK -
1. tried decreasing job to 1 - that didn't work (same error)

2. switched to larger RAM instance (G5.4XL 64GB) to exclude RAM issues, installation still fails, see stack trace below.

3. Is there a recommend image/docker/instance type for running xformers?

5. not sure if related when I run:

`
export CUDA_HOME=/usr/local/cuda-11.4
conda install -c nvidia/label/cuda-11.4.0 cuda-nvcc
conda install -c conda-forge gcc
conda install -c conda-forge gxx_linux-64=9.5.0
`
I get : 

PackagesNotFoundError: The following packages are missing from the target environment:
  - magma-cuda116==2.6
  - fastai==2.1
  - cmake==3.22


Thanks for helping! 






` ninja: build stopped: subcommand failed.
  Traceback (most recent call last):
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 1808, in _run_ninja_build
      subprocess.run(
    File "/opt/conda/envs/pytorch/lib/python3.9/subprocess.py", line 528, in run
      raise CalledProcessError(retcode, process.args,
  subprocess.CalledProcessError: Command '['ninja', '-v']' returned non-zero exit status 1.

  The above exception was the direct cause of the following exception:

  Traceback (most recent call last):
    File "<string>", line 2, in <module>
    File "<pip-setuptools-caller>", line 34, in <module>
    File "/tmp/pip-install-ce15k1tu/xformers_451171c1a51d4c26bb33e63eb012eb41/setup.py", line 264, in <module>
      setuptools.setup(
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/__init__.py", line 87, in setup
      return distutils.core.setup(**attrs)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 185, in setup
      return run_commands(dist)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 201, in run_commands
      dist.run_commands()
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 968, in run_commands
      self.run_command(cmd)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/dist.py", line 1217, in run_command
      super().run_command(command)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 987, in run_command
      cmd_obj.run()
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/command/install.py", line 68, in run
      return orig.install.run(self)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/command/install.py", line 698, in run
      self.run_command('build')
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/cmd.py", line 319, in run_command
      self.distribution.run_command(command)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/dist.py", line 1217, in run_command
      super().run_command(command)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 987, in run_command
      cmd_obj.run()
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/command/build.py", line 132, in run
      self.run_command(cmd_name)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/cmd.py", line 319, in run_command
      self.distribution.run_command(command)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/dist.py", line 1217, in run_command
      super().run_command(command)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 987, in run_command
      cmd_obj.run()
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/command/build_ext.py", line 84, in run
      _build_ext.run(self)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/command/build_ext.py", line 346, in run
      self.build_extensions()
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 765, in build_extensions
      build_ext.build_extensions(self)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/command/build_ext.py", line 466, in build_extensions
      self._build_extensions_serial()
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/command/build_ext.py", line 492, in _build_extensions_serial
      self.build_extension(ext)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/command/build_ext.py", line 246, in build_extension
      _build_ext.build_extension(self, ext)
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/setuptools/_distutils/command/build_ext.py", line 547, in build_extension
      objects = self.compiler.compile(
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 586, in unix_wrap_ninja_compile
      _write_ninja_file_and_compile_objects(
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 1487, in _write_ninja_file_and_compile_objects
      _run_ninja_build(
    File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 1824, in _run_ninja_build
      raise RuntimeError(message) from e
  RuntimeError: Error compiling objects for extension
  error: subprocess-exited-with-error
  
  × Running setup.py install for xformers did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /opt/conda/envs/pytorch/bin/python3.9 -u -c '
  exec(compile('"'"''"'"''"'"'
  # This is <pip-setuptools-caller> -- a caller that pip uses to run setup.py
  #
  # - It imports setuptools before invoking setup.py, to enable projects that directly
  #   import from `distutils.core` to work with newer packaging standards.
  # - It provides a clear error message when setuptools is not installed.
  # - It sets `sys.argv[0]` to the underlying `setup.py`, when invoking `setup.py` so
  #   setuptools doesn'"'"'t think the script is `-c`. This avoids the following warning:
  #     manifest_maker: standard file '"'"'-c'"'"' not found".
  # - It generates a shim setup.py, for handling setup.cfg-only projects.
  import os, sys, tokenize
  
  try:
      import setuptools
  except ImportError as error:
      print(
          "ERROR: Can not execute `setup.py` since setuptools is not available in "
          "the build environment.",
          file=sys.stderr,
      )
      sys.exit(1)
  
  __file__ = %r
  sys.argv[0] = __file__
  
  if os.path.exists(__file__):
      filename = __file__
      with tokenize.open(__file__) as f:
          setup_py_code = f.read()
  else:
      filename = "<auto-generated setuptools caller>"
      setup_py_code = "from setuptools import setup; setup()"
  
  exec(compile(setup_py_code, filename, "exec"))
  '"'"''"'"''"'"' % ('"'"'/tmp/pip-install-ce15k1tu/xformers_451171c1a51d4c26bb33e63eb012eb41/setup.py'"'"',), "<pip-setuptools-caller>", "exec"))' install --record /tmp/pip-record-vvm8klt3/install-record.txt --single-version-externally-managed --compile --install-headers /opt/conda/envs/pytorch/include/python3.9/xformers
  cwd: /tmp/pip-install-ce15k1tu/xformers_451171c1a51d4c26bb33e63eb012eb41/
  Running setup.py install for xformers ... error
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> xformers`

**Comment by danthe3rd** [comment]
The recommended way of installing xformers is using the [conda packages](https://github.com/facebookresearch/xformers#installing-xformers) we provide. No build is necessary then.
It just requires to have pytorch already installed with conda (see here: https://pytorch.org/), but something like this should work:

> conda install pytorch torchvision torchaudio pytorch-cuda=11.7 -c pytorch -c nvidia

**Comment by eran-sefirot** [comment]
that command was successful, is there an easy way to check the status of the installation?

**Comment by eran-sefirot** [comment]
Thanks! 
first let me say thanks for think incredible optimization  - on A10 GPU we were able to increase processing speed from 8it/s to 20it/s cutting our inference time by half.

I have taken some instruction from AITemplate (https://github.com/facebookincubator/AITemplate#clone-the-code)
not sure why but the solved my problems.

To run the optimization I had to:
1) switch from linux to ubuntu.
2) build the docker image 
run
python setup.py bdist_wheel
pip install dist/*.whl --force-reinstall

and then followed the rest of xformers instruction.


Good luck
